### PR TITLE
[feat] add intervention methods for multi-target learning

### PR DESCRIPTION
**Pull request by Dave-AdamsWANG**

### CODE REVIEW

File: tzrec/modules/intervention.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import torch
+from torch import nn
+
+class RotateLayer(nn.Module):
```
Comment (tiankongdeguiji) [inline tzrec/modules/intervention.py:16]: add unit test for these module in intervention_test.py

File: tzrec/models/dc2vr.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+from tzrec.modules.mlp import MLP
+from tzrec.modules.mmoe import MMoE as MMoEModule
+from tzrec.modules.intervention import Intervention 
+from tzrec.protos.model_pb2 import ModelConfig
+from tzrec.protos.models import multi_task_rank_pb2
+from tzrec.utils.config_util import config_to_kwargs
+
+
+
+class DC2VR(MultiTaskRank):
```
Comment (tiankongdeguiji) [inline tzrec/models/dc2vr.py:28]: add model test in dc2vr_test.py

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/modules/intervention.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        self.weight = torch.nn.Parameter(torch.empty(base_dim, low_rank_dim), requires_grad=True)
+        torch.nn.init.orthogonal_(self.weight)
+
+    def forward(self, base):
+        return torch.matmul(base.to(self.weight.dtype), self.weight)
+
+class Intervention(nn.Module):
+    def __init__(self,
+        base_dim,source_dim,
+        low_rank_dim,orth=True):
```
Comment (tiankongdeguiji) [inline tzrec/modules/intervention.py]: the `orth` param has no effect?

File: tzrec/modules/intervention.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import torch
+from torch import nn
+
+class RotateLayer(nn.Module):
+    def __init__(self, 
+                base_dim,
+                low_rank_dim):
+        super().__init__()
+        # n > m
```
Comment (tiankongdeguiji) [inline tzrec/modules/intervention.py:26]: base_dim must greater than low_rank_dim?
```
assert base_dim > low_rank_dim
```

File: tzrec/protos/tower.proto [hunk 5]
```diff
@@ -58,6 +58,29 @@ message BayesTaskTower {
...
+    optional uint32 num_class = 5 [default = 1];
+    // task specific mlp
+    optional MLP mlp = 6;
+    // training loss weights
+    optional float weight = 7 [default = 1.0];
+
+    // intervention tower names
+    repeated string intervention_tower_names = 8;
+    // low_rank_dim
+    required uint32 low_rank_dim = 9;
```
Comment (tiankongdeguiji) [inline tzrec/protos/tower.proto:80]: Is it possible for us to assign a distinct low_rank_dim to each tower?

File: tzrec/modules/intervention.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self,
+        base_dim,source_dim,
+        low_rank_dim,orth=True):
+        super().__init__()
+        self.base_dim = base_dim
+        base_rotate_layer = RotateLayer(base_dim, low_rank_dim)
+        self.base_rotate_layer = torch.nn.utils.parametrizations.orthogonal(base_rotate_layer)
+        source_rotate_layer = RotateLayer(source_dim, low_rank_dim)
+        self.source_rotate_layer = torch.nn.utils.parametrizations.orthogonal(source_rotate_layer)
+        self.dropout = torch.nn.Dropout(0.1)
```
Comment (tiankongdeguiji) [inline tzrec/modules/intervention.py]: dropout rate can be configurable?

**Review by tiankongdeguiji — CHANGES_REQUESTED** [review]

File: tzrec/modules/intervention.py (hunk 1 above)
Comment (Dave-AdamsWANG) [reply tzrec/modules/intervention.py:16]: Added intervention_test.py in modules/

File: tzrec/models/dc2vr.py (hunk 2 above)
Comment (Dave-AdamsWANG) [reply tzrec/models/dc2vr.py:28]: Added  dc2vr_test.py in models/

File: tzrec/modules/intervention.py (hunk 3 above)
Comment (Dave-AdamsWANG) [reply tzrec/modules/intervention.py]: Deleted

File: tzrec/modules/intervention.py (hunk 4 above)
Comment (Dave-AdamsWANG) [reply tzrec/modules/intervention.py:26]: Added the assertion

File: tzrec/protos/tower.proto (hunk 5 above)
Comment (Dave-AdamsWANG) [reply tzrec/protos/tower.proto:80]: All towers are concatenated and projected into a single interaction space in current version (bugs for source dimension in the last version).

File: tzrec/modules/intervention.py (hunk 6 above)
Comment (Dave-AdamsWANG) [reply tzrec/modules/intervention.py]: Fixed in the latest version

**Review by tiankongdeguiji — APPROVED** [review]
LGTM
